src/extension/infer.rs

```diff
@@ -0,0 +1,855 @@
+                }
+            }
+        }
+        // Seperate loop so that we can assume that a metavariable has been
+        // added for every (Node, Direction) in the graph already.
+        for tgt_node in hugr.nodes() {
+            let sig: &OpType = hugr.get_nodetype(tgt_node).into();
+            // Incoming ports with a dataflow edge
+            for port in hugr
+                .node_inputs(tgt_node)
+                .filter(|src_port| matches!(sig.port_kind(*src_port), Some(EdgeKind::Value(_))))
+            {
+                for (src_node, _) in hugr.linked_ports(tgt_node, port) {
+                    let m_src = self
+                        .extensions
+                        .get(&(src_node, Direction::Outgoing))
+                        .unwrap();
+                    let m_tgt = self
+                        .extensions
+                        .get(&(tgt_node, Direction::Incoming))
+                        .unwrap();
+                    self.add_constraint(*m_src, Constraint::Equal(*m_tgt));
+                }
+            }
+        }
+    }
+
+    /// Try to turn mismatches into `ExtensionError` when possible
+    fn report_mismatch(
+        &self,
+        m1: Meta,
+        m2: Meta,
+        rs1: ExtensionSet,
+        rs2: ExtensionSet,
+    ) -> InferExtensionError {
+        let loc1 = self
+            .extensions
+            .iter()
+            .find(|(_, m)| **m == m1 || self.resolve(**m) == m1)
+            .map(|a| a.0);
+        let loc2 = self
+            .extensions
+            .iter()
+            .find(|(_, m)| **m == m2 || self.resolve(**m) == m2)
+            .map(|a| a.0);
+        let err = if let (Some((node1, dir1)), Some((node2, dir2))) = (loc1, loc2) {
+            // N.B. We're looking for the case where an equality constraint
+            // arose because the two locations are connected by an edge
+
+            // If the directions are the same, they shouldn't be connected
+            // to each other. If the nodes are the same, there's no edge!
+            //
+            // TODO: It's still possible that the equality constraint
+            // arose because one node is a dataflow parent and the other
+            // is one of it's I/O nodes. In that case, the directions could be
+            // the same, and we should try to detect it
+            if dir1 != dir2 && node1 != node2 {
+                let [(src, src_rs), (tgt, tgt_rs)] = if *dir2 == Direction::Incoming {
+                    [(node1, rs1.clone()), (node2, rs2.clone())]
+                } else {
+                    [(node2, rs2.clone()), (node1, rs1.clone())]
+                };
+
+                if src_rs.is_subset(&tgt_rs) {
+                    Some(InferExtensionError::EdgeMismatch(
+                        ExtensionError::TgtExceedsSrcExtensions {
+                            from: *src,
+                            from_offset: Port::new(Direction::Outgoing, 0),
```

Comment:
how do you know the ports are 0?

Reply:
The `0`s are just dummy ports to indicate the connection between two nodes. This isn't doing the work to know which ports connect the nodes

Reply:
makes the error quite misleading no? If the ports aren't important to report maybe this needs a new error that just reports the nodes?